planner/core/find_best_task.go

```diff
@@ -418,6 +442,204 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
+		isPartition:      ds.isPartition,
+		physicalTableID:  ds.physicalTableID,
+	}.Init(ds.ctx)
+	statsTbl := ds.statisticTable
+	if statsTbl.Indices[idx.ID] != nil {
+		is.Hist = &statsTbl.Indices[idx.ID].Histogram
+	}
+	is.initSchema(ds.id, idx, true)
+	rowCount := path.countAfterAccess
+	if prop.ExpectedCnt < ds.stats.RowCount {
+		selectivity := ds.stats.RowCount / path.countAfterAccess
+		rowCount = math.Min(prop.ExpectedCnt/selectivity, rowCount)
+	}
+	// scan cost
+	partialCost += rowCount * scanFactor
+	is.stats = property.NewSimpleStats(rowCount)
+	is.stats.StatsVersion = ds.statisticTable.Version
+	if ds.statisticTable.Pseudo {
+		is.stats.StatsVersion = statistics.PseudoVersion
+	}
+	indexConds := path.indexFilters
+	if indexConds != nil {
+		var selectivity float64
+		// cpu cost for index filter
+		partialCost += rowCount * cpuFactor
+		if path.countAfterAccess > 0 {
+			selectivity = path.countAfterIndex / path.countAfterAccess
+		}
+		count := is.stats.RowCount * selectivity
+		stats := &property.StatsInfo{RowCount: count}
+		indexSel := PhysicalSelection{Conditions: indexConds}.Init(is.ctx, stats)
+		indexSel.SetChildren(is)
+		// network cost
+		partialCost += count * netWorkFactor
+		return indexSel, partialCost, count
+	}
+	return is, partialCost, rowCount
+}
+
+func (ds *DataSource) convertToPartialTableScan(prop *property.PhysicalProperty, path *accessPath) (PhysicalPlan, float64, float64) {
+	var partialCost float64
+	ts := PhysicalTableScan{
+		Table:           ds.tableInfo,
+		Columns:         ds.Columns,
+		TableAsName:     ds.TableAsName,
+		DBName:          ds.DBName,
+		isPartition:     ds.isPartition,
+		physicalTableID: ds.physicalTableID,
+	}.Init(ds.ctx)
+	ts.SetSchema(ds.schema)
+	if ts.Table.PKIsHandle {
+		if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+			if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+				ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+			}
+		}
+	}
+	ts.Ranges = path.ranges
+	ts.AccessCondition, ts.filterCondition = path.accessConds, path.tableFilters
+	rowCount := path.countAfterAccess
+	// Adjust number of rows we actually need to scan if prop.ExpectedCnt is smaller than the count we calculated.
+	if prop.ExpectedCnt < ds.stats.RowCount {
+		count, ok, corr := ds.crossEstimateRowCount(path, prop.ExpectedCnt, false)
+		if ok {
+			rowCount = count
+		} else if corr < 1 {
+			correlationFactor := math.Pow(1-corr, float64(ds.ctx.GetSessionVars().CorrelationExpFactor))
+			selectivity := ds.stats.RowCount / rowCount
+			rowCount = math.Min(prop.ExpectedCnt/selectivity/correlationFactor, rowCount)
+		}
+	}
+	ts.stats = property.NewSimpleStats(rowCount)
+	ts.stats.StatsVersion = ds.statisticTable.Version
+	if ds.statisticTable.Pseudo {
+		ts.stats.StatsVersion = statistics.PseudoVersion
+	}
+	// scan cost
+	partialCost += rowCount * scanFactor
+	// network cost
+	partialCost += rowCount * netWorkFactor
+	return ts, partialCost, rowCount
```

Comment:
Currently, in `accessPathsForConds`, we would abandon the path if `len(path.tableFilters) > 0`, but if the path is table scan, we can actually use it indeed.

For example:
```sql
create table t(id int primary key, a int, b int, key idx_a(a));
explain select id from t where (id > 1 and b > 1) or a > 1;
```
currently, we would not choose index merge for it, while actually we should.